### ISSUE TITLE
yaml aliases/anchors not supported

### ISSUE THREAD

**Issue by Timple** [xqms/rosmon#106]
We use yaml aliases to prevent configuration duplication.
When two nodes use a very similar set of parameters we can setup a parameter file like this:
(The two nodes being: `kalman_filter_odom` and `kalman_filter_map`

```yaml
kalman_filter_odom: &KALMAN_FILTER_ODOM
  world_frame: "odom"

  imu0: imu/data
  imu0_config: [false, false, false,  # X,Y,Z,
                true,  true,  true,   # roll,pitch,yaw,
                false, false, false,  # X`,Y`,Z`,
                true,  true,  true,   # roll`,pitch`,yaw`,
                true,  true,  true]   # X``,Y``,Z``

  odom0: wheels/odom
  odom0_config: [false, false, false,  # X,Y,Z,
                  false, false, false,  # roll,pitch,yaw,
                  true,  true,  false,  # X`,Y`,Z`,
                  false, false, true,   # roll`,pitch`,yaw`,
                  false, false, false]  # X``,Y``,Z``

kalman_filter_map:
  <<: *KALMAN_FILTER_ODOM  # All settings and inputs kalman_filter_odom has and:
  world_frame: "map"

  odom1: absolute/odom
  odom1_config: [true,  true,  true,   # X,Y,Z,
                 false, false, false,  # roll,pitch,yaw,
                 false, false, false,  # X`,Y`,Z`,
                 false, false, false,  # roll`,pitch`,yaw`,
                 false, false, false]  # X``,Y``,Z``
```

This is supported by roslaunch in melodic. But rosmon crashes with the following error:

```
Running as '/rosmon_1574149719933920064'

terminate called after throwing an instance of 'ros::InvalidNameException'
  what():  Character [<] at element [19] is not valid in Graph Resource Name [/kalman_filter_map/<</frequency].  Valid characters are a-z, A-Z, 0-9, / and _.
Aborted (core dumped)
```

**Comment by xqms** [comment]
Hi, thanks for the report! Could you share a full `.launch` file that shows how you include the YAML data?

I'm kind of surprised since `rosmon` uses `yaml-cpp` as the YAML parser, which should be pretty feature-complete. Well, you never know ;-)

**Comment by xqms** [comment]
also I just noticed: Where is `frequency` coming from? I guess the YAML file you tried has more fields?

**Comment by Timple** [comment]
[param.launch.txt](https://github.com/xqms/rosmon/files/3864251/param.launch.txt)
[robot_localization.yaml.txt](https://github.com/xqms/rosmon/files/3864252/robot_localization.yaml.txt)

Here ya go. Sorry for the `.txt` extension, seemed most convenient to attach to issue.

I tried to create a minimal example.
These example files work with `roslaunch` but not with `mon launch`

**Comment by xqms** [comment]
Thanks! I had a look and I think the syntax you are using is not supported by yaml-cpp. See here for a PR against yaml-cpp: https://github.com/jbeder/yaml-cpp/pull/300

Furthermore, the syntax is not even in the YAML spec: https://stackoverflow.com/a/47203224

Maybe it is possible to work around this in rosmon, but I doubt it, so we will have to wait for the yaml-cpp developers to implement this.

But there is an alternative: You can do the same at launch file level. Example:
```xml
<node ...>
  <rosparam command="load" file=".../generic.yaml">
  <rosparam command="load" file=".../special_A.yaml">
</node>

<node ...>
  <rosparam command="load" file=".../generic.yaml">
  <rosparam command="load" file=".../special_B.yaml">
</node>
```

rosmon guarantees that `<rosparam>` tags are executed in order. Because the `<rosparam>` tags are scoped inside the node tag, the contents will be loaded inside the node namespace.

**Comment by Timple** [comment]
That must be a discrepancy with the [python yaml import](https://github.com/ros/ros_comm/blob/5da095d06bccbea708394b399215d8a066797266/tools/roslaunch/src/roslaunch/loader.py#L413) then.
That's a bummer...

Thanks for the alternative though! I'll see if we want to convert the files (this will place some of the substitution in two places, param & launch) or switch back to roslaunch.

Implementing a workaround in rosmon sounds like the least neat solution. And since your referenced PR is open since 2015 I have little hopes there...

**Comment by xqms** [comment]
> Implementing a workaround in rosmon sounds like the least neat solution. And since your referenced PR is open since 2015 I have little hopes there...

That could actually be quite easy - just do a first pass over the YAML nodes to find all `<<` nodes and work on these, then the rest (overwriting already set parameters). If you are interested in working on this, I'd be happy to consider a PR - otherwise I might have time to look at this next week.

**Comment by Timple** [comment]
Hmm, might still be an option then to include this. Would be consistent with roslaunch :+1: 
Working towards a deadline I won't get to this again before the holidays :(
If your up to speed with the code, feel free to check next week. Otherwise I might pick it up later!

**Comment by xqms** [comment]
Fixed by @Cartoonman in #117.
